tests/conftest.py

```diff
@@ -95,3 +96,30 @@ def pytest_terminal_summary(terminalreporter: TerminalReporter, exitstatus: Exit
         writer.write("\n### Failures\n\n")
         for failure in failures:
             _write_report(writer, failure)
+
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_runtest_protocol(item):
+    ihook = item.ihook
+    ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
+    reports = threaded_run_report(item)
+    for rep in reports:
+        ihook.pytest_runtest_logreport(report=rep)
+    ihook.pytest_runtest_logfinish(nodeid=item.nodeid, location=item.location)
+    return True
+
+
+def threaded_run_report(item) -> List[TestReport]:
+    # for now, we run setup/teardown in the subprocess
+    # XXX optionally allow sharing of setup/teardown
+    from _pytest.runner import runtestprotocol
+
+    def run_threaded(reports: List[TestReport]):
+        reports.extend(runtestprotocol(item, log=False))
+
+    reports = []
+    thread = threading.Thread(target=run_threaded, args=(reports,))
```

Comment:
Running in a thread will not help regarding memory leak, as it stays shared with main process.
To be sure, I ran `pdm run pytest tests/ -m "memory_expensive"` before and after, and it did not change anything (memory usage goes up to 7go, you can see it locally by `checking _memory_tests.csv`)

You should use subprocess and not threading, so that way memory is actually cleared.

You can do that either using a process pool executor 
https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor-example
especially instiating one and using `max_tasks_per_child=1`.
That way, task would be executed in a separated process and cleaned up

Or you could spawn manually a process and handle, but it will be much more difficult.

Reply:
Honestly, we could kind of replace xdist at the same time, by handling with several pools and tasks, to handle parallelism as we want. But that's another question :)

Reply:
The issue is that `item`, is not pickable. Therefore it's not that easy to start a subprocess, while the fork method works because we don't need to pickle

Reply:
What is the type of item ? maybe we can find some way to serialize/unserialize it ?

Reply:
> It's a Node containing `_Pyhtest.python.Function` which is very hard to serialize because it contains references to a bunch of objects which are not simple to serialize, like the FixtureManager.

https://github.com/pytest-dev/pytest-xdist/issues/675#issuecomment-868400332